Fixed missing Root notifications via geyser plugin framework

lijunwangs opened the pull request:
#### Problem

It is reported that Geyser is missing some Root notifications for slots. https://github.com/solana-labs/solana/issues/31124
The Root notification is sent from replay_stage's code in handle_votable_bank.  https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L1981. However, the validator does not necessarily vote on every slot on the rooted chain. From @carllin 

For instance if the rooted chain is 1->2->3->4

You might only vote on 1 and 4
But when 4 is rooted, 2 is also rooted
But handle_votavle_bank is not called on 2

As result of this, we may miss notifications for slot 2 and 3.

#### Summary of Changes
1. Enhanced BankNotification to add NewRootedChain enum to send the chains of parent roots.
2. Renamed BankNotification::Root -> BankNotification::NewRootBank
3. Introduced SlotNotification for SlotStatusObserver interfaces to send slot status without Bank.
4. In the OptimisticallyConfirmedBankTracker notify parents of a new root if these parents were not notified.
5. Modified and added unit test cases to verify the logic.

Fixes #
<!-- OPTIONAL: Feature Gate Issue: # -->
<!-- Don't forget to add the "feature-gate" label -->

t-nelson commented on rpc/src/optimistically_confirmed_bank_tracker.rs:
```diff
@@ -191,6 +195,29 @@ impl OptimisticallyConfirmedBankTracker {
+        if bank_notification_subscribers.is_none() {
+            return;
+        }
+        for root_bank in bank.clone().parents_inclusive().iter().rev() {
```
i'm not sure `Bank::parents_inclusive()` is what we want to use here. it's going to be a slow linked-list walk to build the `Vec` that it returns and iirc, what's available is limited to what's in the local node's ledger.  `Bank::ancestors` is probably what we want instead.

i think we should also `take_while` from current root back to the last reported root, reverse that, then notify forward, to reduce considering very old roots that have almost certainly been notified already

lijunwangs replied:
Thanks! The code in line 180 uses the same logic. Should we do the same optimization?

lijunwangs replied:
Unfortunately -- we cannot use Bank::ancestors -- the notification interfaces requires a Bank for the ancestor

CriesofCarrots replied:
Do we have any assurances that the ancestor roots haven't already been dropped from BankForks?

lijunwangs replied:
Great question. Looking into it!

carllin replied:
`ancestors` is not safe to use because there could have been new roots made by the time this is called since this is async from replay_stage where the roots are set.

What we probably want is to send a `Vec<Arc<Bank>>` from `replay_stage` where the root is being set. In `set_root` We are already building a vector of parents of the current root, which should include all the parents up until the last root https://github.com/AshwinSekar/solana/blob/2efd72b1a1a06a89613232d92f75a8e1c1995a6a/core/src/replay_stage.rs#L2023-L2024. We could just grab those and send them over this channel.

A more detailed questions though is why do we need the actual banks in these notifications? I'm assuming to access AccountsDB? But having the bank `N` does not guarantee AccountsDB is in the same view as it was during bank `N` because future roots `> N` could have been made and cleaned up state

t-nelson replied:
> the notification interfaces requires a Bank for the ancestor

doesn't look like this is true? i think we only pull the slot and parent slot out of the bank when forwarding the notification. so could extract those earlier and only forward them along?

lijunwangs replied:
> To Tyera's point `ancestors()` is not safe to use because there could have been new roots made by the time this is called since this is async from `replay_stage` where the roots are set, and then earlier rooted banks would have been removed from `BankForks`
> 
> What we probably want is to send a `Vec<Arc<Bank>>` from `replay_stage` where the root is being set. In `set_root` We are already building a vector of parents of the current root, which should include all the parents up until the last root https://github.com/AshwinSekar/solana/blob/2efd72b1a1a06a89613232d92f75a8e1c1995a6a/core/src/replay_stage.rs#L2023-L2024. We could just grab those and send them over this channel.
> 
> A more detailed questions though is why do we need the actual banks in these notifications? I'm assuming to access AccountsDB? But having the bank `N` does not guarantee AccountsDB is in the same view as it was during bank `N` because future roots `> N` could have been made and cleaned up state

Thanks @carllin -- as pointed by @t-nelson, we can change the notifications by sending the (slot, parent_slot)  pair without requiring the bank. Regarding the race condition on the roots on the BankForks -- I wonder why https://github.com/AshwinSekar/solana/blob/2efd72b1a1a06a89613232d92f75a8e1c1995a6a/rpc/src/optimistically_confirmed_bank_tracker.rs#L176 does not have the same concern?

carllin replied:
I think the same issue is present there as well :)

t-nelson replied:
i guess we could reduce the scope of this pr to "be consistently broken", rather than uniquely broken? :slightly_smiling_face:

CriesofCarrots replied:
Or fix the other place first...?

lijunwangs replied:
> Or fix the other place first...?

I will fix this issue first and the investigate the other next to reduce the scope.

carllin commented on rpc/src/optimistically_confirmed_bank_tracker.rs:
```diff
@@ -45,14 +72,23 @@ impl std::fmt::Debug for BankNotification {
 pub type BankNotificationSender = Sender<BankNotification>;
 
+#[derive(Clone)]
+pub struct BankNotificationSenderOption {
```
rename to `BankNotificationSenderConfig` since the purpose of this struct is to be a configuration to toggle whether or not to `send_parents` depending if `geyser` is enabled

lijunwangs replied:
Done

carllin commented on rpc/src/optimistically_confirmed_bank_tracker.rs:
```diff
@@ -45,14 +72,23 @@ impl std::fmt::Debug for BankNotification {
+#[derive(Clone)]
+pub struct BankNotificationSenderOption {
+    pub sender: BankNotificationSender,
+    pub send_parents: bool,
```
`send_parents` -> `should_send_parents`

lijunwangs replied:
Done

carllin commented on rpc/src/optimistically_confirmed_bank_tracker.rs:
```diff
@@ -35,7 +35,34 @@ impl OptimisticallyConfirmedBank {
-    Root(Arc<Bank>),
+    /// Send the Bank currently rooted and optionally the parent slots in (Slot, Parent Slot) pairs
+    /// The parent slots contains the slot of the bank.
+    Root((Arc<Bank>, Option<Vec<(Slot, Slot)>>)),
```
Instead of a tuple I think we should make two separate enums, can be two separate notifications

1. `NewRootBank(Arc<Bank>)`
2. `NewRootedChain(Vec<Slot>)`, the parents here are inferred by the chain, we do not need the tuple

lijunwangs replied:
The reason for sending the parents pair can can send the parent for the highest slot in the chain.

carllin commented on rpc/src/optimistically_confirmed_bank_tracker.rs:
```diff
@@ -191,6 +231,27 @@ impl OptimisticallyConfirmedBankTracker {
         }
     }
 
+    fn notify_root_banks(
```
`notify_root_banks -> notify_new_root_slots`

lijunwangs replied:
Done

carllin commented on rpc/src/optimistically_confirmed_bank_tracker.rs:
```diff
@@ -102,7 +140,8 @@ impl OptimisticallyConfirmedBankTracker {
         highest_confirmed_slot: &mut Slot,
-        bank_notification_subscribers: &Option<Arc<RwLock<Vec<BankNotificationSender>>>>,
+        highest_root_slot: &mut Slot,
+        bank_notification_subscribers: &Option<Arc<RwLock<Vec<SlotNotificationSender>>>>,
```
nit: rename this and all the other references to `bank_notification_subscribers` to `slot_notification_subscribers`

lijunwangs replied:
Done

carllin commented on rpc/src/optimistically_confirmed_bank_tracker.rs:
```diff
@@ -35,7 +35,34 @@ impl OptimisticallyConfirmedBank {
+    Root((Slot, Slot)),
+}
+
+impl From<BankNotification> for SlotNotification {
+    fn from(notification: BankNotification) -> Self {
+        match notification {
+            BankNotification::OptimisticallyConfirmed(slot) => {
+                SlotNotification::OptimisticallyConfirmed(slot)
+            }
+            BankNotification::Frozen(bank) => {
+                SlotNotification::Frozen((bank.slot(), bank.parent_slot()))
+            }
+            BankNotification::Root(bank) => {
+                SlotNotification::Root((bank.0.slot(), bank.0.parent_slot()))
+            }
+        }
+    }
```
Is this implementation necessary? Seem like it's not being used anywhere

lijunwangs replied:
Removed

carllin commented on rpc/src/optimistically_confirmed_bank_tracker.rs:
```diff
@@ -191,6 +231,27 @@ impl OptimisticallyConfirmedBankTracker {
 
+    fn notify_root_banks(
+        roots: &mut Vec<(Slot, Slot)>,
+        slot_threshold: Slot,
```
I feel like this could be `newest_root: &mut Slot` and we just update it to the newest root after `roots.sort_unstable();` below

lijunwangs replied:
Done

carllin commented:
For clarity in the future, big changes like the switches from `BankNotification` -> `SlotNotification` could be pulled into a separate PR to make reviewing easier

CriesofCarrots reviewed (commented):
I think I'm happy with the changes. However, optimistically_confirmed_bank_tracker is getting really difficult to grok. Can you update the PR summary of changes to include more details about the intermediate changes? eg. why are there two different notification types and sender/receivers defined?
Can you also update the module docs as well?

> For clarity in the future, dense changes like the switches from BankNotification -> SlotNotification could be pulled into a separate PR to make reviewing easier

Yes, also this.

lijunwangs commented:
> For clarity in the future, dense changes like the switches from `BankNotification` -> `SlotNotification` could be pulled into a separate PR to make reviewing easier

The reason for that is it would be more messy to use BankNotification in the SlotStatusObserver interfaces with the changed BankNotification interface.

CriesofCarrots reviewed (dismissed):
Lgtm with Trent's nit

t-nelson reviewed (approved):
:shipit:

lijunwangs commented:
Enhanced BankNotification to add NewRootedChain enum to send the chains of parent roots.
Renamed BankNotification::Root -> BankNotification::NewRootBank
Introduced SlotNotification for SlotStatusObserver interfaces to send slot status without Bank.
In the OptimisticallyConfirmedBankTracker notify parents of a new root if these parents were not notified.
Modified and added unit test cases to verify the logic.

fanatid commented:
Do you have a plan to make a backport to 1.13/1.14?

t-nelson commented:
Definitely _not_ 1.13. 1.14 bp probably unavoidable
